lambda-buffers-compiler/test/Test/DeriveCheck.hs

```diff
@@ -0,0 +1,627 @@
+
+import A
+
+data Foo a = Foo (Maybe a)
+
+instance C a => C (Foo a)
+-}
+moduleB'3 :: ModuleBuilder
+moduleB'3 =
+  ModuleBuilder
+    { mbTyDefs = tyDefsB
+    , mbInstances = instancesB
+    , mbClasses = S.empty
+    , mbScope = scopeB
+    }
+  where
+    tyDefsB :: M.Map Text Exp
+    tyDefsB =
+      M.fromList
+        [ "Foo" .= DecE (LitE (Name "Foo")) (LitE (TyVar "a") *: nil) $
+            mkSum
+              [ "Foo" .= toProdE [ForeignRefE ["A"] "Maybe" :@@ LitE (TyVar "a")]
+              ]
+        ]
+
+    scopeB =
+      S.fromList
+        [ C _c (ForeignRefP ["A"] "Int") :<= []
+        , C _c (ForeignRefP ["A"] "Bool") :<= []
+        , C _c (ForeignRefP ["A"] "String") :<= []
+        , C _c (ForeignRefP ["A"] "Maybe" :@ _X) :<= [C _c _X]
+        ]
+
+    instancesB = S.fromList [C _c (LocalRefP "Foo" :@ _X) :<= [C _c (LocalRefP "Foo" :@ _X)]]
+
+{- Test 5. Overlapping instances. Should fail w/ Overlap error
+
+-- imports are as if this were A
+module A where
+
+(... same as above ...)
+instance C (Maybe Int)
+
+-----------------------------
+
+module B where
+
+data Foo = Foo (Maybe Int)
+
+instance C Foo
+
+-}
+moduleB'4 :: ModuleBuilder
+moduleB'4 =
+  ModuleBuilder
+    { mbTyDefs = tyDefsB
+    , mbInstances = instancesB
+    , mbClasses = S.empty
+    , mbScope = scopeB
+    }
+  where
+    tyDefsB :: M.Map Text Exp
+    tyDefsB =
+      M.fromList
+        [ "Foo" .= DecE (LitE (Name "Foo")) NilE $
+            mkSum
+              [ "Foo" .= toProdE [ForeignRefE ["A"] "Maybe" :@@ ForeignRefE ["A"] "Int"]
+              ]
+        ]
+
+    scopeB =
+      S.fromList
+        [ C _c (ForeignRefP ["A"] "Int") :<= []
+        , C _c (ForeignRefP ["A"] "Bool") :<= []
+        , C _c (ForeignRefP ["A"] "String") :<= []
+        , C _c (ForeignRefP ["A"] "Maybe" :@ _X) :<= [C _c _X]
+        , C _c (ForeignRefP ["A"] "Maybe" :@ ForeignRefP ["A"] "Int") :<= []
+        ]
+
+    instancesB = S.fromList [C _c (LocalRefP "Foo") :<= []]
+
+{- Test 6. Should fail w/ missing Int instance
+
+same as moduleB'2 w/ the imported (C Int) instance removed
+
+-}
+moduleB'5 :: ModuleBuilder
+moduleB'5 =
+  ModuleBuilder
+    { mbTyDefs = tyDefsB
+    , mbInstances = instancesB
+    , mbClasses = S.empty
+    , mbScope = scopeB
+    }
+  where
+    tyDefsB :: M.Map Text Exp
+    tyDefsB =
+      M.fromList
+        [ "Foo" .= DecE (LitE (Name "Foo")) NilE $
+            mkSum
+              [ "Foo" .= toProdE [ForeignRefE ["A"] "Maybe" :@@ ForeignRefE ["A"] "Int"]
+              ]
+        ]
+
+    scopeB =
+      S.fromList
+        [ C _c (ForeignRefP ["A"] "Bool") :<= []
+        , C _c (ForeignRefP ["A"] "String") :<= []
+        , C _c (ForeignRefP ["A"] "Maybe" :@ _X) :<= [C _c _X]
+        ]
+
+    instancesB = S.fromList [C _c (LocalRefP "Foo") :<= []]
+
+{- Test 7. Should fail w/ constructor in constraint context error.
+
+same as moduleB'2 except the instance dec is replaced by:
+
+instance C (Maybe Int) => C Foo
```

Comment:
Hmm I didn't know that was a problem. Can you elaborate a bit?

Reply:
SPJ says this is necessary for well-behaved instances in a non-MPTC/"extensionless" context. He doesn't elaborate in the paper and I couldn't track down an explanation elsewhere (maybe it's in an ancient version of the Haskell report?). I believe it has to do w/ termination of the constraint solver.